src/dialer/dial-request.js

```diff
@@ -0,0 +1,197 @@
+'use strict'
+
+const AbortController = require('abort-controller')
+const AggregateError = require('aggregate-error')
+const pDefer = require('p-defer')
+const debug = require('debug')
+const log = debug('libp2p:dialer:request')
+log.error = debug('libp2p:dialer:request:error')
+const { AbortError } = require('libp2p-interfaces/src/transport/errors')
+
+const { anySignal } = require('../util')
+const { TokenHolder } = require('./token-holder')
+
+class DialRequest {
+  /**
+   *
+   * @param {object} options
+   * @param {Multiaddr[]} options.addrs
+   * @param {TransportManager} options.transportManager
+   * @param {Dialer} options.dialer
+   */
+  constructor ({
+    addrs,
+    dialAction,
+    dialer
+  }) {
+    this.addrs = addrs
+    this.dialer = dialer
+    this.dialAction = dialAction
+  }
+
+  /**
+   * @async
+   * @param {object} options
+   * @param {AbortSignal} options.signal An AbortController signal
+   * @param {number} options.timeout The max dial time for each request
+   * @returns {Connection}
+   */
+  async run (options) {
+    // Determine how many tokens we need
+    const tokensWanted = Math.min(this.addrs.length, this.dialer.perPeerLimit)
+    // Get the tokens
+    const tokens = this.dialer.getTokens(tokensWanted)
+    // If no tokens are available, throw
+    if (tokens.length < 1) {
```

Comment:
Right now we don't have a mechanism for this, other than observing `dialer.token.length`. I went with a throw approach when there are no tokens to simplify the dial logic. If a dial can happen right now, it will be excited, otherwise it just fails immediately. It does push retry/queue logic onto the user, if they wish to queue up dials. This does also have an impact on things like DHT queries. If we have enough queries to consume all available tokens, dial attempts for new queries will start to fail.

We could alleviate this by making `getTokens` a generator and allowing token dials to stack up until tokens are freed. This will create problems with queue saturation (which we already have in master). One thing I would like to eventually look into is adding Connection "prepurchasing" to Connection Management. This would enable subsystems to request a pool of connections prior to performing certain actions, such as a DHT query. Subsystems could then take a proactive approach to ratelimiting their actions.

The DHT and initial startup are the most common scenarios that would likely cause us to hit token limits within the system itself. Ideally, we should be avoiding scenarios where we're attempting to establish `MAX_PARALLEL_DIALS` all at once.